NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp

```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
+public:
+	explicit CRoundRobinBuffer( IMathEngine& mathEngine, int numItems, int itemSize ) :
+		temp( mathEngine, /*bytes-size*/numItems * itemSize ), raw( NeoML::GetRaw( temp.GetHandle() ) ),
+		numItems( numItems ), itemSize( itemSize ), currPos( 0 ),
+		prevBatchedSourceRowStart( -1 ), prevNumItemsToAdd( 0 ), startItem( 0 )
+	{}
 
-		int inputRow = inputRowStart;
-		for(int filterRow = filterRowBackStart;
-			filterRow >= filterRowBackEnd;
-			filterRow -= desc.StrideHeight, ++inputRow) {
-			// The temp blob stores the filter rows multiplied by input; add them to the output rows in correct positions
-			const float* tempRowData = tempRaw + ( ( batch * input.Height() + inputRow )
-				* input.Width() * filter.Height() + filterRow ) * filter.Width() * filterChannels;
-
-			for(int col = -desc.PaddingWidth;
-				col <= output.Width() + desc.PaddingWidth - filter.Width();
-				col += desc.StrideWidth) {
-
-				int tempRowDataShift = 0;
-				int toCopy = filter.Width();
-				int pos = col;
-				if(pos < 0) {
-					tempRowDataShift = -pos;
-					toCopy += pos;
-					pos = 0;
-				}
-				if(pos + toCopy > output.Width()) {
-					toCopy = output.Width() - pos;
-				}
-				if( toCopy > 0 ) {
-					toCopy *= filterChannels;
-					tempRowDataShift *= filterChannels;
-					float* outputVec = outputDataPtr + pos * filterChannels;
-					vectorAdd( outputVec, tempRowData + tempRowDataShift, outputVec, toCopy );
-				}
-				tempRowData += filter.Height() * filter.Width() * filterChannels;
-			}
-		}
+	int GetExistItemsNum( int batchedSourceRowStart, int numItemsToAdd )
+	{
+		const int numItemsExist = std::max( 0, prevBatchedSourceRowStart + prevNumItemsToAdd - batchedSourceRowStart );
+		startItem = numItemsExist ? ( ( startItem + batchedSourceRowStart - prevBatchedSourceRowStart ) % numItems ) : 0;
+		prevBatchedSourceRowStart = numItemsToAdd ? batchedSourceRowStart : prevBatchedSourceRowStart;
+		prevNumItemsToAdd = numItemsToAdd;
+		// It is forbidden to divide appending matrix lines in two chunks (in the buffer's beginning and end)!
+		ASSERT_EXPR( numItemsExist == 0 || ( numItemsToAdd - numItemsExist ) <= ( numItems - currPos ) );
+		return numItemsExist;
 	}
-}
+	int GetSourceRowNum( int sourceRow ) const { return ( sourceRow + startItem ) % numItems; }
+	float* GetBufPtr( int numItemsExist, int numItemsToAdd )
+	{
+		float* const ptr = raw + ( numItemsExist ? currPos : 0 ) * itemSize; //if there are no lines to keep, start buffer from the scratch
+		currPos = numItemsExist ? ( ( currPos + numItemsToAdd - numItemsExist ) % numItems ) : 0;
+		return ptr;
+	}
+	const float* GetRaw( int offset ) const { return raw + offset; }
+	int GetPrevBatchedSourceRowStart() const { return prevBatchedSourceRowStart; }
+
+private:
+	CFloatHandleStackVar temp;
+	float* const raw;
+	const int numItems, itemSize;
+	int currPos /*start possition to add the buffer*/, prevBatchedSourceRowStart, prevNumItemsToAdd, startItem /*start of usefull data in buffer*/;
```

Comment:
Each field must be on a separate line.
And each field must be commented.

Reply:
fixed